### ISSUE TITLE
php updater/application.php upgrade:detect fails

### ISSUE THREAD

**Issue by joostdekeijzer** [owncloud/updater#482]
Hi,

Using Owncloud 10.0.9.5 (just updated)

Running `php updater/application.php upgrade:detect` fails with the following error (shortened):

> The command "php occ --no-warnings config:list --output=json '--private='" failed. 
>  
>  Error Output:                                                                                                                   
>                                                                                                                                  
>    The "--private" option does not accept a value.                                                                               
>                                                                                                                                  
>  config:list [--output [OUTPUT]] [--private] [--] [<app>]                                                                        

In `Owncloud\Updater\Utils\ConfigReader::load()` the `--private` option is added to the args of `Owncloud\Updater\Utils\OccRunner::runJson()` with an empty value.

In `Owncloud\Updater\Utils\OccRunner::run()` the args array is parsed, resulting in the command:

`php occ --no-warnings config:list --output=json '--private='`

Maybe, for empty values, the run command should skip the "=", resulting in `php occ --no-warnings config:list --output=json '--private'` in this case? (The single quotes around the argument are ok).

I can create a PR but have no idea of the consequences of this change in other parts of owncloud...

**Comment by danieldjewell** [comment]
Goededag @joostdekeijzer!

You were absolutely correct about the "=" --  the ultimate problem appears to stem from here: https://github.com/owncloud/updater/commit/f950190acb52efb34145414fb2250354c2d0fd6c 

Combined with the use of an empty string as the "value" for the $args associative array as I referenced in #485 - this results in the argument being "--private=" instead of "--private" 

It would appear that the design of OccRunner::run (at 69) specifically takes into account this situation and tests whether the key of the array starts with "--" ... it even uses the "===" operator to test if the "--" is not found. If it is not found, it doesn't use an "=".

This would be consistent with designing OccRunner::run() to expect associative arrays (with the key set to "--key") for --key=value pairs and simple array elements for simple --arg parameters. 

https://github.com/owncloud/updater/blob/master/src/Utils/OccRunner.php#L73-L80

Consider:

```php
$args = array('--output' => 'json', '--private', '--something', '--hello', '--debug' => 'true');
print_r($args);
```
Output: 

```
Array
(
    [--output] => json
    [0] => --private
    [1] => --something
    [2] => --hello
    [--debug] => true
)
```

In this case, https://github.com/owncloud/updater/blob/master/src/Utils/OccRunner.php#L74 would test "FALSE" for all but the first and last elements and would fall-through to https://github.com/owncloud/updater/blob/master/src/Utils/OccRunner.php#L77. 

So I guess the question becomes, is the design of OccRunner::run() not correct or is the offending code in my pull #485 the problem? I believe it's the code in ``ConfigReader::runJson():109`` that's the issue. But it all depends on whether or not ``run()'s`` parsing code is working as intended.

**Comment by danieldjewell** [comment]
Please also see owncloud/updater#482 for more notes

**Comment by PVince81** [comment]
@VicDeo

**Comment by danieldjewell** [comment]
I think this should be resolved against [master] (PR #485) and [stable10] (PR #500) ... Not sure if there are other branches that need this.

**Comment by VicDeo** [comment]
Should be fixed with https://github.com/owncloud/updater/pull/500
please reopen if it is not
